repo/blob/gdrive/gdrive_storage.go

```diff
@@ -0,0 +1,460 @@
+// Package gdrive implements Storage based on Google Drive.
+package gdrive
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/google"
+	drive "google.golang.org/api/drive/v3"
+	"google.golang.org/api/googleapi"
+	"google.golang.org/api/option"
+
+	"github.com/kopia/kopia/internal/clock"
+	"github.com/kopia/kopia/internal/iocopy"
+	"github.com/kopia/kopia/internal/retry"
+	"github.com/kopia/kopia/repo/blob"
+	"github.com/kopia/kopia/repo/blob/retrying"
+)
+
+const (
+	gdriveStorageType = "gdrive"
+	gdriveMimeType    = "application/x-kopia"
+	
+	uploadChunkSize   = 1 << 20
+	uploadContentType = "application/octet-stream"
+	maxPrefixLen      = 26
+	queryRetryDelay   = 100 * time.Millisecond
+	queryRetryMax     = 3
+
+  // googleapi.Field values.
+	listIdFields       = "files(name,id)"
+	listMetadataFields = "files(name,id,mimeType,size,modifiedTime)"
+	getMetadataFields  = "name,id,mimeType,size,modifiedTime"
+)
+
+type gdriveStorage struct {
+	Options
+
+	client      *drive.Service
+	folderId    string
+	// Caches blob id -> drive file id associations.
+	fileIdCache map[blob.ID]string
+}
+
+func (gdrive *gdriveStorage) GetBlob(ctx context.Context, b blob.ID, offset, length int64, output blob.OutputBuffer) error {
+	if offset < 0 {
+		return blob.ErrInvalidRange
+	}
+
+	fileId, err := gdrive.GetFileId(ctx, b)
+	if err != nil {
+		return err
```

Comment:
in general prefer `errors.Wrap()` over `return err` - this helps with tracking down sources of errors.

Reply:
Thanks for the tip. I went through and added Wrapf to all the places where we interface with drive api.